### DISCUSSION TITLE
需要完善文档，补充add_from_source选项的介绍

### DISCUSSION THREAD

**Discussion by zhongfly** [Tony15246/uosc_danmaku#57] [General]
我发现add_from_source选项没有在文档中记录，并且功能也相当实用，我个人理解是打开网络视频时，会添加网络视频来源的弹幕，例如播放b站的链接会加载B站的弹幕。但实际并不是，应该可以实现吧？

**Comment by dyphire** [comment]
实现b站网络视频加载弹幕的用法并不难，简单处理下 url 就行。只是之前没人提这个需求（

**Reply by zhongfly** [reply]
目前看来，这个功能更适合放在autoload_for_url…

**Reply by dyphire** [reply]
大概看了下，之前想简单了，直接从 path 解析 url 的话光 b 站的 url 播放链接就好几种格式，会员内容甚至不传 BV 号，不好处理
通过 ytdlp 播放还好处理一些，通过 Play-With-MPV 播放更难搞了。前者不支持巴哈，如果只做 b 站适配的话倒简单了

**Reply by zhongfly** [reply]
就直接传path的url呗，用弹弹play的接口。没获取到的就算了
与yt-dlp合作的话，可以做成从字幕轨中获取，一般名称会是danmaku。

**Comment by Tony15246** [comment]
`add_from_source`其实主要是为本地视频开发的，作用是关联过一次的弹幕源会被记录，并且下次播放同一个视频的时候会自动关联添加过的弹幕源。

> 目前看来，这个功能更适合放在autoload_for_url

确实如此

**Reply by Tony15246** [reply]
b站的视频我记得唯一标识符应该是cid，只要拿到cid就能通过`https://comments.bilibili.com/{$cid}.xml`的形式从官方接口获取弹幕。

如果用户是通过yt-dlp播放的b站视频，那yt-dlp本身就会塞一个弹幕轨道，参考下面这个方式就能拿到cid

https://github.com/itKelis/MPV-Play-BiliBili-Comments/blob/ad7b801ff93b1d825ba83876228c694659e42b64/scripts/bilibiliAssert/main.lua#L104-L111

如果用户是通过 Play-With-MPV 播放，Play-With-MPV 本身就会把cid传给脚本，直接获取就行。就算Play-With-MPV不传cid，也是有办法fallback地获取的，参考我和一个在linux上使用Play-With-MPV的难兄难弟的讨论

https://github.com/LuckyPuppy514/Play-With-MPV/issues/149#issuecomment-2338319866

**Comment by dyphire** [comment]
在 #58 中实现为 bilibili 视频播放加载弹幕，本地测试无问题，可以尝试看看

**Reply by dyphire** [reply]
Play-With-MPV 支持的巴哈的弹幕加载等好心人实现吧（

**Reply by Tony15246** [reply]
巴哈的情况比b站更纯粹一些，yt-dlp不支持直接解析并播放巴哈的链接，用mpv看巴哈大概率是通过Play-With-MPV 播放的。其实不管是不是通过Play-With-MPV 播放的，巴哈的url都只有一种格式，就是这样的`https://bahamut.akamaized.net/某路径/720p/身份验证相关信息,用户名%3a38881%3a一些乱七八糟的校验/chunklist_b1200000.m3u8`。其中被两个`%3a`也就是冒号包裹的就是巴哈的sn号，只要拿到sn号，和b站一样就能通过官方的弹幕接口`https://ani.gamer.com.tw/ajax/danmuGet.php`获取弹幕。基本上应该可以参考这个仓库的实现。

https://github.com/s594569321/MPV-Play-BAHA-Comments

**Reply by Tony15246** [reply]
等b站的pr处理完我可以实现一下

**Reply by dyphire** [reply]
巴哈弹幕集成不是我不知道如何处理，是我无法在巴哈网页看到 Play-With-MPV 脚本正常工作。无法调试的功能不会去做

**Reply by Tony15246** [reply]
我和我的同学本地巴哈网页 Play-With-MPV 脚本都可以正常工作，应该可以涵盖linux和windows环境下的调试测试。

Play-With-MPV 脚本无法正常工作的情况下，应该也可以通过用

~`mpv "https://bahamut.akamaized.net/某路径/720p/身份验证相关信息,用户名%3a38881%3a一些乱七八糟的校验/chunklist_b1200000.m3u8" --http-header-fields="origin: https://ani.gamer.com.tw/" `~

的方式来播放巴哈的视频并测试。

这个命令好像有点问题，我记得稍加修改就不会403来着

对了，这么写就行`mpv "https://bahamut.akamaized.net/某路径/720p/身份验证相关信息,用户名%3a38881%3a一些乱七八糟的校验/chunklist_b1200000.m3u8" --http-header-fields="origin: https://ani.gamer.com.tw" `

Play-With-MPV也是通过传递相同的链接播放的，只要在上面这种情况下能正常工作就能兼容Play-With-MPV

**Comment by Tony15246** [comment]
#61 中实现为巴哈姆特视频播放加载弹幕，本地测试无问题

**Reply by Tony15246** [reply]
经测试windows系统下也无问题
